Wait device ops before starting the profile

jonb377 opened the pull request:
Tracing can get ahead of device execution for Llama2 training because there are no transfers to/from device. This means that the tracing step can be inaccurate when triggered at a certain step in Python, since device execution may lag behind.

This change waits for device ops to finish before triggering the profile.

jonb377 committed:
Wait device ops before starting the profile

alanwaketan reviewed (approved):
It LGTM. Should we comment it that this is going to slow down the training for a bit?

BTW, what's the use case in your mind about this change? Like you figure some bugs are harder to debug because of the de-sync between tracing and executions?

jonb377 commented:
> BTW, what's the use case in your mind about this change?

For multislice tests, sometimes the host/slice which is taking the capture can get ahead of other slices, so the profile is taken before device execution actually starts (some straggler is holding up training). 

We can workaround by taking the profile at a later step, but this is more robust - I've found that the execution queue has different sizes on different hardware, so tracing will get further ahead on some hardware generations and require a later profile step.

jonb377 committed:
Add comment and move capture to the end of step

jonb377 commented:
> Should we comment it that this is going to slow down the training for a bit?

Thanks @alanwaketan, that's a good point - I've moved it after tracing to minimize the impact and added a comment.
